mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs

```diff
@@ -353,248 +427,293 @@ impl RingBuffer {
 
 impl Drop for RingBuffer {
     fn drop(&mut self) {
-        let result = self.mmap.flush();
+        let result = self.file.flush();
         if let Some(err) = result.err() {
             error!("Failed to flush {:?}", err);
         }
+        let result = self.metadata_file.flush();
+        if let Some(err) = result.err() {
+            error!("Failed to flush metadata {:?}", err);
+        }
     }
 }
 
-fn create_file(file_path: &Path) -> IOResult<File> {
+fn create_file(file_path: &PathBuf) -> IOResult<File> {
+    if let Some(parent) = file_path.parent() {
+        create_dir_all(parent)?;
+    }
     OpenOptions::new()
         .read(true)
         .write(true)
         .create(true)
         .open(file_path)
 }
 
+fn save_ring_buffer_metadata(metadata: &RingBufferMetadata, file: &mut File) -> BincodeResult<()> {
+    let buf = bincode::serialize(metadata)?;
+    file.seek(SeekFrom::Start(0))?;
+    file.write_all(&buf)?;
+    file.flush()?;
```

Comment:
We always flush for metadata, as the alternative of having to scan the whole file (especially if large) felt unacceptable to me.

Reply:
can metadata follow the same flushing strategy? In normal circumstances the buffer and metadata should always be in sync. The only issue is when process crashes/killed. Two things that should ensure consistency:
- treat metadata as an optimizatiopn but not the source of truth - that's what you're already doing.
- add checksum to metadata, disregard corrupted metadata and fallback to full scan (unfortunately).

What do you think?

Reply:
Yeah, I agree. Lets merge 2 threads https://github.com/Azure/iotedge/pull/4429/files#r577213662

Reply:
Good point, didn't think about it like that. Let me go ahead and make that change too. As mentioned in another thread, it might be possible to get away from needing the metadata file altogether, but I was seeing long times for init. Might be having to deserialize a bunch, but not sure.

Reply:
As I mentioned above: another option is to write metadata only on graceful shutdown, and re-scan the file on crash. Checksum on metadata is good to have in any case.